py-polars/tests/unit/operations/test_replace.py

```diff
@@ -503,3 +478,60 @@ def test_map_dict_deprecated() -> None:
     with pytest.deprecated_call():
         result = s.to_frame().select(pl.col("a").map_dict({2: 100})).to_series()
     assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    ("context", "dtype"),
+    [
+        (pl.StringCache(), pl.Categorical),
+        (pytest.warns(CategoricalRemappingWarning), pl.Categorical),
+        (contextlib.nullcontext(), pl.Enum(["a", "b", "OTHER"])),
+    ],
+)
+def test_replace_cat_str(
+    context: contextlib.AbstractContextManager,  # type: ignore[type-arg]
+    dtype: pl.DataType,
+) -> None:
+    with context:
+        for old, new, expected in [
+            ("a", "c", pl.Series("s", ["c", None], dtype=pl.Utf8)),
+            (["a", "b"], ["c", "d"], pl.Series("s", ["c", "d"], dtype=pl.Utf8)),
+            (pl.lit("a", dtype=dtype), "c", pl.Series("s", ["c", None], dtype=pl.Utf8)),
+            (
+                pl.Series(["a", "b"], dtype=dtype),
+                ["c", "d"],
+                pl.Series("s", ["c", "d"], dtype=pl.Utf8),
+            ),
+        ]:
```

Comment:
You can just add a second `parametrize` decorator to the test to check all the permutations.

Reply:
The values depend on the values in the first parametrize (e.g. `pl.lit("a",dtype=dtype`). Do you know how to do that?

Reply:
It's possible, let me double check...

Reply:
Hm, not trivially, it seems. The other test can be rewritten though with a parametrize like below.

I guess it's fine as it is 👍 

```python
@pytest.mark.parametrize(
    "context", [pl.StringCache(), pytest.warns(CategoricalRemappingWarning)]
)
@pytest.mark.parametrize(
    ("old", "new", "expected"),
    [
        (
            "a",
            pl.lit("c", dtype=pl.Categorical),
            pl.Series("s", ["c", None], dtype=pl.Categorical),
        ),
        (
            ["a", "b"],
            pl.Series(["c", "d"], dtype=pl.Categorical),
            pl.Series("s", ["c", "d"], dtype=pl.Categorical),
        ),
    ],
)
def test_replace_cat_cat(
    context: contextlib.AbstractContextManager,  # type: ignore[type-arg]
    old: str | list[str],
    new: pl.Series | pl.Expr,
    expected: pl.Series,
) -> None:
```

Reply:
The expected still depends on the context, so I'm affraid this won't work

Reply:
You're right!